PEP 728: TypedDict with Typed Extra Fields

PIG208 opened the pull request:
<!--
You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.

If your PEP is not Standards Track, remove the corresponding section.
-->

## Basic requirements (all PEP Types)

* [x] Read and followed [PEP 1](https://peps.python.org/1) & [PEP 12](https://peps.python.org/12)
* [x] File created from the [latest PEP template](https://github.com/python/peps/blob/main/peps/pep-0012/pep-NNNN.rst?plain=1)
* [x] PEP has next available number, & set in filename (``pep-NNNN.rst``), PR title (``PEP 123: <Title of PEP>``) and ``PEP`` header
* [x] Title clearly, accurately and concisely describes the content in 79 characters or less
* [ ] Core dev/PEP editor listed as ``Author`` or ``Sponsor``, and formally confirmed their approval
* [x] ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly
* [x] ``PEP-Delegate``, ``Topic``, ``Requires`` and ``Replaces`` headers completed if appropriate
* [x] Required sections included
    * [x] Abstract (first section)
    * [x] Copyright (last section; exact wording from template required)
* [ ] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer names](https://peps.python.org/pep-0012/#literal-blocks) if non-Python
* [ ] PEP builds with no warnings, pre-commit checks pass and content displays as intended in the rendered HTML
* [x] Authors/sponsor added to ``.github/CODEOWNERS`` for the PEP


## Standards Track requirements

* [x] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) with general agreement that a PEP is appropriate
* [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [x] Motivation
    * [x] Rationale
    * [x] Specification
    * [x] Backwards Compatibility
    * [ ] Security Implications
    * [ ] How to Teach This
    * [x] Reference Implementation
    * [x] Rejected Ideas
    * [x] Open Issues
* [x] ``Python-Version`` set to valid (pre-beta) future Python version, if relevant
* [x] Any project stated in the PEP as supporting/endorsing/benefiting from the PEP formally confirmed such
* [ ] Right before or after initial merging, [PEP discussion thread](https://peps.python.org/pep-0001/#discussing-a-pep) created and linked to in ``Discussions-To`` and ``Post-History``


<!-- readthedocs-preview pep-previews start -->
----
:books: Documentation preview :books:: https://pep-previews--3441.org.readthedocs.build/

<!-- readthedocs-preview pep-previews end -->

PIG208 committed:
PEP 728: TypedDict with Typed Extra Fields

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 commented:
Hi @AlexWaygood! Thanks for expressing interest in this feature. Could you confirm yourself as a sponsor of this PEP?

JelleZijlstra commented:
There is significant overlap with some of the ideas in #3440, a proposed rewrite of PEP 705. It may be worth discussing the interaction of the two ideas on Discuss before committing to a PEP.

PIG208 commented:
While I assumed that literal blocks starting with `::` should have Python syntax highlighting, it does not seem to be the case in the review. Not sure if I'm missing something?

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,614 @@
+Title: TypedDict with Typed Extra Fields
+Author: Zixuan James Li <zixuan.li@nyu.edu>
+Sponsor: Alex Waygood <alex.waygood@gmail.com>
+Discussions-To: https://discuss.python.org/t/pep-692-using-typeddict-for-more-precise-kwargs-typing/17314
```
Should be a new thread, not the one for PEP 692

PIG208 replied:
Before this is merged, should I create a new "Typing" thread or leave it blank until a "pep" thread can be created?

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,614 @@
+        name: str
+        __extra__: bool
+    
+    a: Movie = {"name": "Blade Runner", "novel_adaption": True}  # OK
```
```suggestion
    a: Movie = {"name": "Blade Runner", "novel_adaptation": True}  # OK
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,614 @@
+``__required_keys__`` and ``__optional_keys__`` should include ``__extra__`` in
+runtime.
+
+Because this is a type-checking feature, it can be made available to older
```
Or even without typing_extensions. The PEP as currently written would require few if any changes to the runtime implementation.

PIG208 replied:
Also regarding alternatives to `__extra__`, I think something like this will not require changes to the runtime implementation at all:
```
    class Foo(TypedDict):
        def __extra_field__(self, key: str) -> int:
            ...
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,614 @@
+
+While this saves us from reserving an attribute for special use, it will require
+extra work to implement inheritance, and it is less natural to integrate with
+generics if we support generic TypedDict types in the future.
```
We already support generic TypedDicts, there's no "future" about it.

JelleZijlstra reviewed (commented):
A few comments.

I'll also note that I don't like this proposal's current form of using a special-cased key. I would prefer some mechanism that doesn't look like the TypedDict has a key `"__extra__"`.

alicederyn commented:
Author of the PEP-705 rewrite! Excited to see this :) The syntax I'm going for in PEP-705 is

```python
class Foo(TypedDict, other_keys=False):
```

I'm happy to rewrite the proposal to just reference this PEP. (I have a personal preference for `other_keys=Never` over `__extra__ = Never`.)

My main concern in adding `other_keys` to PEP-705 is that declaring a `TypedDict` as `@final` to indicate other keys won't be present no longer semantically makes sense if keys can be read-only, so we need an officially sanctioned approach.

AlexWaygood commented:
> Hi @AlexWaygood! Thanks for [expressing interest](https://discuss.python.org/t/pep-692-using-typeddict-for-more-precise-kwargs-typing/17314/92) in this feature. Could you confirm yourself as a sponsor of this PEP?

I was assuming you'd send me a draft before making a PR to the peps repo 😄 But no worries!

In principle, I'm still happy to sponsor something along these lines, but please give me some time to read this and PEP-705 over and think about it before confirming. Ping me again if I haven't got back to you in a few days :)

AA-Turner commented:
Converting to draft to mark that there's no sponsor (yet).

A

PIG208 committed:
Address review comments.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

alicederyn commented:
@AlexWaygood Did you get a chance to look at this and PEP-705 yet? I modified 705 to bring it more in line with this draft, and it's now gone out for another round of comments.

alicederyn commented:
Eric gave some [great feedback](https://discuss.python.org/t/pep-705-typeddict-read-only-and-other-keys/36457/6) on the subset-variant of this PEP that I included in PEP-705. pyright has also changed to allow type discrimination without the `@final` decorator. As a result, it no longer seemed appropriate to couple our PEPs, and I have deferred this completely to you.

I had a couple ideas for alternative syntax that might address Eric's points.

## Ellipsis

Option 1: use an ellipsis (`...`) instead of `__extra__`. This works fine in alternative syntax and also I believe in a typescript-like inline syntax?

```python
Foo = TypedDict("Foo", { "bar": int, ...: str })
foo: { "bar": int, ...: str }
```

But it is a SyntaxError in class definition syntax:

```python
class Foo(TypedDict):
  bar: int
  ...: str  # syntax error
```

That leaves us deciding between:
* a language change to allow this, which would need some serious justification,
* this feature only working in alternative syntax, or
* using different syntax for class definitions

I think it looks decent though, especially in inline syntax, so throwing it out there.

## A type qualifier

Option 2: use a type qualifier instead of a "reserved" key name.

```python
class Foo(TypedDict):
  bar: int
  _: OtherValues[str]
```

Users can use any key that isn't already used, so there's no risk of accidentally changing the meaning of existing definitions, but the PEP can recommend a conventional default to use whenever possible. I like underscore as it is short and it is already conventional to use this for "other stuff" but haven't put much thought into it.

This also has the advantage of being hyperlinked to documentation explaining what it is in editors, making it more discoverable than `__extra__`.

PIG208 commented:
I kind of like the type qualifier idea, but it can have issues with subclassing to address. Will revisit the draft probably this weekend and incorporate some new progress made from recent discussions on PEP-705.

erictraut commented:
@PIG208, what is the status of this PEP? Are you looking to move forward with it? It came up in a [recent typing discussion](https://discuss.python.org/t/do-we-want-a-final-behavior-for-exact-typeddict/44418).

PIG208 commented:
I need to catch up with the previous discussions on the topic to get it up-to-date. I just went back from a break and will push some updates soon.

stroxler commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,614 @@
+    class MovieWithYear(MovieBase):  # OK
+        x: NotRequired[int | None]
+
+As a side-effect, annotating ``__extra__`` with :class:`typing.Never` for a
```
It may be worth making this change much louder than as a side-effect.

This is a key feature people have wanted independent of the more general question of typing extra fields, and in fact type checkers have at times used `@final` for this but there's a very reasonable view that `@final` is inappropriate.

I brought up the problem of exact TypedDict in [this typing thread](https://discuss.python.org/t/do-we-want-an-exact-typeddict-if-so-how-final-extras-never/44418/4) and @erictraut pointed me at PEP 728, but it took me a little while to find this aside.

PIG208 replied:
Updated the PEP to extend on this topic.

JelleZijlstra commented:
I talked to @AlexWaygood and I offered to take over as the sponsor of the PEP. If that's OK with you, @PIG208, please add my name to the headers. I'll also do another review pass on the PEP once you've addressed the open feedback.

Feel free to email me (my email address is all over this repo) to discuss the logistics and process of moving the PEP forward.

PIG208 committed:
Update sponsor information.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Extend on the "final" side-effect.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Mention the limitation of the current "__extra__" approach.

I left out the rewrite to other proposed syntax such as using
"OtherFields[]" with an arbitrary key before there is sufficient
discussion on the alternatives available to us.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 commented on peps/pep-0728.rst:
```diff
@@ -216,38 +235,36 @@ function parameters still apply::
 
     * Changing a field type of a parent TypedDict class in a subclass is not allowed.
 
 First, it is not allowed to change the type of ``__extra__`` in a subclass::
```
If PEP 705 is accepted, `ReadOnly[]` on `OtherFields[]` will make it possible to refine the type in a subclass. However, I am unsure if this section is where the comment belongs to.

PIG208 commented:
Thank you all for the feedback! I have addressed the existing review comments so far and left some questions.

The last commit is an experiment to adapt @alicederyn's second suggested alternative. It does come with some additional complexity when two TypedDicts have `OtherFields[]` annotated keys named differently. I'm also open to discussions on finding better alternatives.

I also feel that the proposal as-is might be too verbose, and much of it consists of reiterating PEP 589 in a different context with examples. So future revisions can be made to shrink it down.

PIG208 commented:
Dropping 0ee6e86 (support for the `OtherFields` alternative) and going back to `__extra__` unless we get a better proposal.

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,633 @@
+        reveal_type(movie["year"])  # Revealed type is 'int'
+
+
+Interaction with Totality
```
We'll also need to specify the interaction with PEP 705. It seems to me that `ReadOnly` should be allowed on `__extra__`, with the obvious meaning: all the extra fields are read-only.

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,633 @@
+        year: NotRequired[int | None]
+
+Due to this nature, an important side effect allows us to define a TypedDict
+type that acts similarly to decorating the class definition with ``@final``::
```
The current consensus is that `@final` has no meaning on TypedDict, so we should not mention it here. Instead, this means that additional keys are disallowed.

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,633 @@
+        name: str
+        __extra__: Never
+
+Here, annotating ``__extra__`` with :class:`typing.Never` prevents ``MovieFinal``
```
This isn't necessarily true under PEP 705 (since you could have a read-only item with a type you refine in a subclass). Also, pyright allows this (I think correctly):

```
from typing import Any, TypedDict

class TD1(TypedDict):
    x: list[Any]
    
class TD2(TD1):
    x: list[str]
```

More generally, I don't think "inheritance" is the right concept to talk about here. Because TypedDicts are structural types, inheritance isn't very important: you can get exactly the same type just by copy-pasting all the fields. What `__extra__: Never` does is say that there can be no other items in the TypedDict, whether in a subclass or otherwise.

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,633 @@
+    extra_str = extra_int  # Not OK. 'int' is inconsistent with 'str' for field '__extra__'
+    
+When only one of the TypedDict ``t`` has ``__extra__``, then the keys unknown to
+``t`` in the other TypedDict corresponds to the ``__extra__`` pseudo-field.
```
```suggestion
``t`` in the other TypedDict correspond to the ``__extra__`` pseudo-field.
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,633 @@
+and its structural subtypes will never have any required key during static
+analysis.
+
+If there is no required key, the TypedDict type is consistent with ``Dict[...]``
```
`Dict` is deprecated, use `dict`. Maybe write `dict[KT, VT]` so you have a way to talk about `VT` instead of "the value type".

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,633 @@
+    }
+
+How Else Will This Interact with PEP 692?
+-----------------------------------------
```
It's not clear what the concrete open issue is here.

PIG208 replied:
Got it removed. I think the "Interaction with `Unpack`" section should be sufficient.

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,633 @@
+
+While dunder attributes like ``__extra__`` are reserved for stdlib, it is still
+a limitation that ``__extra__`` is no longer usable as a regular key. If the
+proposal is accepted, neither ``__required_keys__`` and ``__optional_keys__``
```
As well as PEP 705's additions, I think `__readonly_keys__`.

PIG208 committed:
Address some feedback (WIP).

TODOs:

- rewrite the motivation section
- use inline references instead
- extend on how this PEP interacts with PEP 705

PIG208 committed:
Address feedbacks that require some more substantial rewrites.

In this update:

- The Motivation section has been rewritten to expand more on the
  compelling use cases and focus less on the current workarounds.
- Update the "Open Issues" section to remove the issues that have been
  resolved.
- Add an entry to "Rejected Ideas" after removing it from "Open Issues",
  and keep the discussion relevant to Python as opposed to TypeScript.
- Remove the "References" and use inline hyperlinks instead throughout
  the document.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Update inheritance on interaction with PEP 705.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Redo "type consistency" addressing PEP 705 interaction.

Also split parts of it into individual sections dicussing changes to
TypedDict type consistency with Mapping and dict.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,632 @@
+keys are known, it would be useful to explicitly expect additional keys of
+certain value types.
+
+However, :pep:`589` is more restrictive on type checking the construction of a
```
Nit: We should ideally frame this around the current typing spec, which derives from PEP 589 but can be changed independently.

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,632 @@
+    fun({"name": "Blade Runner", "year": 1982})  # Not OK
+
+While the restriction is enforced when constructing a TypedDict, due to
+structural subtyping, the TypedDict is flexible enough to have extra fields that
```
```suggestion
structural subtyping, the TypedDict may have extra fields that
```
More concise wording

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,639 @@
+
+When ``__extra__`` is annotated with ``ReadOnly[]``, the extra fields on the
+TypedDict have the properties of read-only fields. This affects subclassing
+according to the inheritance rules specified in :pep:`PEP 705 <589#Inheritance>`.
```
Wrong link

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,639 @@
+Operations that are available to ``NotRequired`` fields should also be available
+to the extra fields::
+
+    def f(movie: Movie) -> None:
```
Might be better to repeat the definition of the `Movie` class here, since it's some way away.

JelleZijlstra commented on peps/test.py:
```diff
@@ -0,0 +1,12 @@
+from typing import NotRequired, TypedDict, Unpack
```
Don't think this file should go into this repo

JelleZijlstra commented:
The last round of changes makes some great improvements. I left a few quick comments, but will do another review and then hopefully we can merge the PEP in.

One of the next steps should be to implement the PEP in https://github.com/python/typing_extensions. I think that should just require some minor adjustments to `__required_keys__` logic.

PIG208 committed:
Minor wording and example changes.

Addresses some review feedback.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Mention the typing spec instead of PEP 589 more when possible.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Remove an accident.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 commented:
Thank you for the review @JelleZijlstra ! Opened https://github.com/python/typing_extensions/pull/329.

PIG208 commented:
I'm considering renaming `__extra__` to `__extra_items__` to reduce possible naming collisions and use `item` instead of `field` for consistency with PEP 705 and the typing spec.

PIG208 committed:
Be consistent with "key", "item", "value type" and word choices.

This follows the style of PEP 705 and the majority of the typing spec to
use "key" or "item" instead of "field" depending on the context. It also
makes use of "value type" more to be adequately more pedantic. The
special key added is now referred to as the "reserved key" to make it
clearer that it is not intended for other purposes anymore.

It also make use of type aliases like `VT` more for brevity.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+========
+
+This PEP proposes a way to type extra items for :class:`~typing.TypedDict` using
+a reserved ``__extra__`` key. This addresses the need to define a subset of
```
Noting for the record that I'm not completely happy with this syntax, because it feels wrong to pick one specific key to special-case. However, other possible solutions also have problems, and I'm OK with merging the PEP as is; the PEP discussion can cover possible alternatives.

If we do stick with this version, I'd prefer the shorter name `__extra__` over `__extra_keys__`. The longer name doesn't really make a collision much less likely.

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+Motivation
+==========
+
+A TypedDict type can annotate the value type of each known item from
+dictionaries. However, due to structural subtyping, the TypedDict can have extra
+items that are not visible through its type. There is currently not a way to
+limit or restrict the value type of items that might be present in the TypedDict
+type's structural subtype.
```
```suggestion
A :py:data:`typing.TypedDict` type can annotate the value type of each known item in
a dictionary. However, due to structural subtyping, a TypedDict can have extra
items that are not visible through its type. There is currently no way to
restrict the types of items that might be present in the TypedDict
type's structural subtypes.
```
Editing

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+Defining a Closed TypedDict Type
+--------------------------------
+
+In practice, this prevents users from defining a closed TypedDict type
+when it is expected that the TypedDict type contains no additional items.
```
```suggestion
The current behavior of TypedDict prevents users from defining a closed TypedDict type
when it is expected that the type contains no additional items.
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+            reveal_type(entry)  # Revealed type is 'Movie | Book'
+
+Nothing prevents a ``dict`` that is structurally compatible with ``Movie`` to
+have the ``author`` key, and it would be correct for the type checker to not 
+narrow its type.
```
```suggestion
have the ``author`` key, and under the current specification it would be incorrect for the type checker to 
narrow its type.
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+certain value types.
+
+However, the typing spec is more restrictive on type checking the construction of a
+TypedDict, `preventing the users <https://github.com/python/mypy/issues/4617>`__
```
```suggestion
TypedDict, `preventing users <https://github.com/python/mypy/issues/4617>`__
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+    fun({"name": "Blade Runner", "year": 1982})  # Not OK
+
+While the restriction is enforced when constructing a TypedDict, due to
+structural subtyping, the TypedDict is may have extra items that are not
```
```suggestion
structural subtyping, the TypedDict may have extra items that are not
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+    fun(movie)  # OK
+
+It is not possible to acknowledge the existence of the extra items through
+``in`` check and access them without breaking type safety, even though they
```
```suggestion
``in`` checks and access them without breaking type safety, even though they
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+Using TypedDict Types
+---------------------
+
+For a TypedDict type that has the "__extra__" key, during construction, the
```
```suggestion
For a TypedDict type that has the ``__extra__`` key, during construction, the
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+    }  
+
+In this example, ``__extra__: bool`` does not mean that ``Movie`` has a required
+string key "__extra__" whose value type is ``bool``. Instead, it specifies that
```
```suggestion
string key ``"__extra__"`` whose value type is ``bool``. Instead, it specifies that
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+according to the inheritance rules specified in :pep:`PEP 705 <705#Inheritance>`.
+
+Notably, a subclass of the TypedDict type may redeclare the value type of
+``__extra__`` or extra items if the TypedDict type declares ``__extra__`` to be
```
```suggestion
``__extra__`` or of additional non-extra items if the TypedDict type declares ``__extra__`` to be
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+    movie: Movie = details  # Not OK. While 'int' is consistent with 'int | None',
+                            # 'int | None' is not consistent with 'int'
+
+The value type of "year" is not found in ``Movie``, and that it is not
```
I don't follow this sentence. It might be enough to just have the examples and skip this paragraph of text. (Rename the second `MovieDetails` to something else.)

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+    movie: Movie = details  # Not OK. 'year' is not required in 'Movie',
+                            # so it shouldn't be required in 'MovieDetails' either
+
+Because "year" is absent in ``Movie``, "__extra__" is considered the
```
```suggestion
Because "year" is absent in ``Movie``, ``__extra__`` is considered the
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+                            # so it shouldn't be required in 'MovieDetails' either
+
+Because "year" is absent in ``Movie``, "__extra__" is considered the
+corresponding key. "year" being required violates the rule "For each required
```
```suggestion
corresponding key. ``"year"`` being required violates the rule "For each required
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+is an item defined in ``Movie``.
+
+``__extra__`` as a pseudo-item follows the same rules that other items have,
+so when both TypedDict contains ``__extra__``, this check is naturally
```
```suggestion
so when both TypedDicts contain ``__extra__``, this check is naturally
```

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+        reveal_type(movie.values())  # Revealed type is 'dict_values[str, str]'
+
+Interaction with dict[KT, VT]
+--------------------------------
```
```suggestion
-----------------------------
```
(This is what's breaking the CI checks.)

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,647 @@
+Acknowledgments
+================
+
+Thanks to Jelle Zijlstra for sponsoring this PEP and providing review feedbacks,
```
```suggestion
Thanks to Jelle Zijlstra for sponsoring this PEP and providing review feedback,
```

PIG208 committed:
Apply edits from Jelle.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Another grammatical pass.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
More heading alignment fixes.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 commented:
Hmm, not sure why is `.../peps/pep-0728.rst:29: WARNING: py:data reference target not found: typing.TypedDict`

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,642 @@
+Motivation
+==========
+
+A :py:data:`typing.TypedDict` type can annotate the value type of each known
```
```suggestion
A :py:class:`typing.TypedDict` type can annotate the value type of each known
```
Sorry for this, it's documented as a class not a data

JelleZijlstra committed:
Update peps/pep-0728.rst

JelleZijlstra committed:
Merge branch 'main' into pep-728

JelleZijlstra committed:
Remove Discussions-To for now, can re-add after it's merged

JelleZijlstra commented:
Pushing a couple of small fixes to satisfy the merge requirements, I think we're almost there!

JelleZijlstra commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,640 @@
+Allowing Extra Items without Specifying the Type
+------------------------------------------------
+
+``extra=True`` was originally proposed for defining a TypedDict accept extra
```
```suggestion
``extra=True`` was originally proposed for defining a TypedDict that accepts extra
```

JelleZijlstra committed:
Update peps/pep-0728.rst

JelleZijlstra commented:
And it's merged! It should be up at https://peps.python.org/pep-0728 momentarily. As a next step, could you post in https://discuss.python.org/c/peps/19 about the PEP, and then update the Discussions-To/Post-History headers here in another PR.

PIG208 commented:
Got it, thanks!

hugovk commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,640 @@
+Python-Version: 3.13
+
+
+.. highlight:: rst
```
Please remove this, the code blocks should default to Python instead.

PIG208 replied:
Was wondering about why that was happening. Thanks for spotting that!

hugovk commented on peps/pep-0728.rst:
```diff
@@ -0,0 +1,640 @@
+
+Thanks to Jelle Zijlstra for sponsoring this PEP and providing review feedback,
+Eric Traut who `proposed the original design
+<https://mail.python.org/archives/list/typing-sig@python.org/message/3Z72OQWVTOVS6UYUUCCII2UZN56PV5II/>`
```
This isn't rendering as a link properly, I think it needs a  couple of underscores at the end?

hugovk reviewed (commented):
Please could you add the Discourse link to Discussions-To and Post-History headers?

Edit: done in https://github.com/python/peps/pull/3658 ✅
